pkg/routine/lock.go

```diff
@@ -0,0 +1,220 @@
+)
+
+var (
+	ErrCouldNotLock    = errors.New("lock could not be obtained")
+	ErrCouldNotRelease = errors.New("lock could not be released")
+)
+
+type Lock struct {
+	Name string
+
+	locker  *redislock.Client
+	lockTTL time.Duration
+
+	lock  *redislock.Lock
+	mutex sync.Mutex
+}
+
+type LockOption func(l *Lock)
+
+func NewLock(name string, opts ...LockOption) *Lock {
+	initClient()
+	l := &Lock{Name: name}
+	for _, opt := range []LockOption{ // default options
+		SetTTL(5 * time.Second),
+		SetClient(getDefaultLocker()),
+	} {
+		opt(l)
+	}
+	for _, opt := range opts {
+		opt(l)
+	}
+	return l
+}
+
+func (l *Lock) Acquire(ctx context.Context) (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	opts := &redislock.Options{
+		RetryStrategy: redislock.NoRetry(),
+		Context:       ctx,
+	}
+
+	lock, err := l.locker.Obtain(l.Name, l.lockTTL, opts)
+	if err != nil {
+		log.Ctx(ctx).Debug().Err(err).Str("lockName", l.Name).Msg("Could not acquire lock")
+		switch {
+		case errors.Is(err, redislock.ErrNotObtained):
+			return false, nil
+		default:
+			return false, pberrors.Hide(ctx, err, ErrCouldNotLock)
+		}
+	}
+
+	l.lock = lock
+	return true, nil
+}
+
+func (l *Lock) AcquireWait(ctx context.Context) error {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	opts := &redislock.Options{
+		RetryStrategy: redislock.LinearBackoff(1 * time.Second),
+		Context:       ctx,
+	}
+
+	lock, err := l.locker.Obtain(l.Name, l.lockTTL, opts)
+	if err != nil {
+		log.Ctx(ctx).Debug().Err(err).Str("lockName", l.Name).Msg("Could not acquire lock")
+		return pberrors.Hide(ctx, err, ErrCouldNotLock)
+	}
+
+	l.lock = lock
+	return nil
+}
+
+// AcquireAndKeepUp will acquire a lock, and keep it up constantly until cancel is called,
+// the returned context is a lock context and is detached from the parent context, meaning that
+// any cancellation/timeout on the parent context does not affect this lock context.
+func (l *Lock) AcquireAndKeepUp(ctx context.Context) (context.Context, context.CancelFunc, error) {
+	opts := &redislock.Options{
+		RetryStrategy: redislock.NoRetry(),
+		Context:       ctx,
+	}
+
+	lock, err := l.locker.Obtain(l.Name, l.lockTTL, opts)
+	if err != nil {
+		switch {
+		case errors.Is(err, redislock.ErrNotObtained):
+			return nil, nil, nil
+		default:
+			return nil, nil, pberrors.Hide(ctx, err, ErrCouldNotLock)
+		}
+	}
+
+	// Keep up lock, cancel lockCtx otherwise.
+	// To pass the detached context from the goroutine we need to pass it through a channel in the parent goroutine
+	lockCtxChan := make(chan context.Context)
+	cancel := Run(ctx, func(ctx context.Context) {
+		defer pberrors.HandleWithCtx(ctx, fmt.Sprintf("keep up lock %q", l.Name)) // handle panics
+
+		lockCtxChan <- ctx
+
+		keepUpLock(ctx, lock, l.lockTTL)
+		err := lock.Release()
+		if err != nil && err != redislock.ErrLockNotHeld {
+			log.Ctx(ctx).Debug().Err(err).Msgf("could not release lock %q", l.Name)
+		}
+	})
+
+	return <-lockCtxChan, cancel, nil
```

Comment:
The only thing preventing the "lock" logic to sit in its own package is the dependency on `routine.Run` to start this goroutine.

The biggest reason for using `routine.Run` is when the caller does not care about canceling the goroutine, but we still want to know within that goroutine when it is canceled (because the entire service is being shut down: it received a `syscall.SIGTERM`)

Since `AcquireAndKeepUp` controls the started goroutine (and also gives control to the caller to cancel it), we can simply replace it with the following:

```go
// Keep up lock, cancel lockCtx otherwise.
lockCtx, cancelLock := context.WithCancel(ctx)
go func() {
	defer pberrors.HandleWithCtx(ctx, fmt.Sprintf("keep up lock %q", l.Name)) // handle panics
	defer cancelLock()

	keepUpLock(lockCtx, lock, l.lockTTL)
	err := lock.Release()
	if err != nil && err != redislock.ErrLockNotHeld {
		log.Ctx(ctx).Debug().Err(err).Msgf("could not release lock %q", l.Name)
	}
}()

return lockCtx, cancelLock, nil
```

This gets rid of the circular dependency and allows us to put the lock logic in its supposed package.

Reply:
@mihailozarinschi @arnold-iakab can this be resolved?

Reply:
Yes, the suggested changes were used and tested.